DOC/RF: Lots of docstrings; some `Item`-ization

aqw opened the pull request:
This was an accidental adevnture while I was working on something else. I was motivated for a time to improve docstrings, and so i ran with it while I had the motivation.

When I work on docstrings, I usually do a review (of varying depth) of the code it covers. Thus, there are a few code changes sprinkled in here as well.

And then a few more functions in `inventory.py` have been `Item`-ized.

TobiasKadelka commented on onyo/lib/git.py:
```diff
@@ -354,65 +329,63 @@ def get_hexsha(self,
-        -------
-        list of Path
-          Paths in `paths` that are excluded by the patterns in `ignore`.
+            Paths to check
```
Normally we still have a dot at the end of these lines.

aqw replied:
Thanks for the review! I have addressed most issues.

I responded to the `Returns` section question. I feel fairly strongly about the path we have chosen (and is the path chosen by other projects).

The periods I just don't care enough about.

TobiasKadelka commented on onyo/lib/git.py:
```diff
@@ -354,65 +329,63 @@ def get_hexsha(self,
+        Parameters
+        ----------
+        lines
+            List of text from ``git log`` of a single commit.
```
I think I know what this explanation means, but it is the least clear parameter explanation I read so far, maybe it can be descripted more clearly.

Is each element in the list one line of the `git log` output, including some for empty lines, some for the operation record and so on?

aqw replied:
I have tweaked the wording here a bit, but it's not worth much effort as it is a private function that is only ever used in the context of someone who is actively altering `GitRepo.history()`.

Indeed, your understanding is correct. It is a list of lines of text for one commit that is output from `git log`.

TobiasKadelka commented on onyo/lib/git.py:
```diff
@@ -450,24 +423,28 @@ def _parse_log_output(self, lines: list[str]) -> dict:
-          Limit history going back ``n`` commits.
-          ``None`` for no limit (default).
+        path
+            The Path to get the history of. Defaults to ``HEAD`` (default).
```
Now this sounds as if "HEAD" was a path like every other.

I know explaining this clearly is tricky, but I still want to highlight that I don't fully get it and feel like I have to guess.

aqw replied:
Yeah. I bumped on that too. Fixed.

aqw commented on onyo/lib/command_utils.py:
```diff
@@ -141,18 +135,18 @@ def inline_path_diff(source: str | Path,
                     seq2: Sequence) -> Tuple[int | None, int | None]:
-    """
-    Compare two sequences. If a match is found, return both indexes of the match.
+    r"""Get the indexes of the next matching value of two sequences.
```
Good catch. Fixed. :-)

aqw commented on onyo/lib/git.py:
```diff
@@ -157,33 +151,27 @@ def get_subtrees(self,
-        -------
-        bool
-          True if the git worktree is clean, otherwise False.
+        r"""Whether the git worktree is clean.
```
Yeah, the thing is that having a "Returns" section suppresses the auto generation of a Returns section from the type annotations.

So we either:
1) define everything twice, and have Returns be unlike all other documentation
2) don't define Returns, and describe in the description if there anything interesting about the contents of what is returned (beyond the already-generated type).

We have been pursuing the latter for awhile now, and IMO is the better choice.

TobiasKadelka replied:
Okay, thanks for clarifying!
That makes sense, I just wasn't aware of the change :)

aqw commented on onyo/lib/command_utils.py:
```diff
@@ -141,18 +135,18 @@ def inline_path_diff(source: str | Path,
     ----------
     seq1
-        First sequence.
+        First sequence
```
Oh, and for what it's worth, it's not defined in numpy no enforced by `ruff`.

aqw commented on onyo/lib/command_utils.py:
```diff
@@ -27,48 +27,42 @@
 
-    Renaming (i.e. changing the last element) is its own action, and does not
-    group with other changes. Moves group with adjacent changes when possible.
+    A rename (i.e. changing the last element) is its own action, and does not
```
I have put "move" and "rename" in bold.

aqw commented:
Thanks for the review! I have addressed most of your comments.

I responded to the Returns section question. I feel fairly strongly about the path we have chosen (and is the path chosen by other projects).

The periods I just don't care enough about.
